### DISCUSSION TITLE
issue: error disappears on blur with async validation

### DISCUSSION THREAD

**Discussion by rtkac** [react-hook-form/react-hook-form#7858] [Q&A] [answered]
### Version Number

7.27.0

### stackblitz demo

https://stackblitz.com/edit/react-yxqusm

### Steps to reproduce

1. Go to [stackblitz demo](https://stackblitz.com/edit/react-yxqusm)
3. Type `1111` into Code input field (there is a 1s delay on purpose)
4. Click outside of the input field
5. The error `ASYNC field is wrong` under the field disappears
6. Also check the console output under the app window
7. You can see that the form is valid and there are no errors:
`code hasErrors =>  false`
`Form isValid => true`

Code field has async validation (validation code [here](https://stackblitz.com/edit/react-yxqusm?file=src%2FformValidator.js)).

### Expected behaviour

The error under the field should persist on blur event.

### More information

- `mode: 'all'` is used with useForm
- There is [vest](https://github.com/ealush/vest) resolver used (vest@4.1.3)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
vest resolver will based on what's return by the vest validation, you can debug in the below code snippet:

```tsx
import React from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import { vestResolver } from '@hookform/resolvers/vest';
import { DevTool } from '@hookform/devtools';

import { formValidator } from './formValidator';

import { Input } from './Input';

export default function App() {
  const formFunctions = useForm({
    mode: 'all',
    resolver: async (data, context, options) => {
      console.log('result', await vestResolver(formValidator())(data, context, options))
      return await vestResolver(formValidator())(data, context, options)
    },
    defaultValues: {
      code: '',
    },
  });

  const {
    control,
    handleSubmit,
    formState: { errors, isValid },
  } = formFunctions;

  const onSubmit = (data) => {
    console.log('Form submit data: ', JSON.stringify(data));
  };

  // console.log('Form isValid => ', isValid);
  // console.log('Form errors => ', errors);

  return (
    <div className="App">
      <FormProvider {...formFunctions}>
        <form onSubmit={handleSubmit(onSubmit)}>
          <Input name="code" label="Code" />

          <button type="submit">Submit</button>
        </form>
        <DevTool control={control} placement="top-right" />
      </FormProvider>
    </div>
  );
}
```

my observation, after blur this is what is returned by the resolver, if you need to check code at the resolver level: here it is https://github.com/react-hook-form/resolvers/blob/master/vest/src/vest.ts thanks to Vest author putting this together.

<img width="753" alt="Screen Shot 2022-02-19 at 10 05 06 am" src="https://user-images.githubusercontent.com/10513364/154772707-4d41df84-031d-4c88-9fbb-3a2addf9d73b.png">

**Comment by rtkac** [comment]
@bluebill1049 thanks for your quick reply. Do you have any thoughts how to solve this?

**Comment by Moshyfawn** [comment]
I assume, it's the validation logic error and not the RHF's. If you pay attention to what the `skipWhen` return value is `onBlur`, it's `true`.
Your `formValue.code === previousCode` is `true` `onBlur`. Your `previousCode` value is `undefined` until `onBlur`. `onBlur` `previousCode` becomes the same as `formValue.code`. As a result, `skipWhen` returns `true`.

Now, as per the description of  [`skipWhen` from the Vest docs](https://vestjs.dev/docs/writing_your_suite/including_and_excluding/skipWhen)
> When using skipWhen the tests within the block will be skipped, but will still be counted against isValid

So, my bet is it's smth to do with Vest logic, 'cuz the RHF resolver isn't suppose to alter any of the validation lib logic behavior, it's just an abstraction.

**Comment by rtkac** [comment]
@Moshyfawn thanks, I thing you're right. Here is working [demo](https://stackblitz.com/edit/react-dl5vey) (in case of someone has similar issue).

**Reply by Moshyfawn** [reply]
Yep, looks good in your latest example!
So, you just check `(res) => res.hasErrors('code')` instead of a custom field value condition as in your previous example, correct?

**Reply by rtkac** [reply]
I think that `res.hasErrors('code')` has no effect on that behaviour. In my opinion is that the whole test function is written different.
 `
test.memo(
   'code',
  'Code already taken',
  () => {
    return validateCode(data.code);
  },
  [data.code]
);
`

`
async function validateCode(code) {
  await wait(1000);
  return enforce(code).notEquals('1111');
}
`

**Reply by Moshyfawn** [reply]
Yeah, I was just checking out [Vest's docs on `.memo`](https://vestjs.dev/docs/writing_tests/advanced_test_features/test.memo#testmemo-for-memoized-tests).
Such an interesting and feature-rich lib! Had it in my bookmarks but wasn't going to try it out until your examples.

**Reply by rtkac** [reply]
You should try it! Test cases could be written really easy.
